mutexbot/src/main.rs

```diff
@@ -0,0 +1,251 @@
+        self.failure_count += 1;
+        info!("Failed to send request! Retrying in 2 seconds...");
+        sleep(Duration::from_secs(2)).await;
+        Ok(())
+    }
+    async fn status_code(&mut self, status_code: StatusCode) -> Result<()> {
+        if !status_code.is_server_error() || self.failure_count >= 15 {
+            anyhow::bail!("Unexpected status code: {status_code}")
+        }
+        self.failure_count += 1;
+        info!("Server error, status code {status_code}. Retrying in 2 seconds...");
+        sleep(Duration::from_secs(2)).await;
+        Ok(())
+    }
+}
+
+#[derive(Serialize)]
+struct ReservePayload {
+    notes: String,
+    duration: Option<String>,
+    isolation_channel: Option<String>,
+}
+
+#[derive(Serialize)]
+struct ReleasePayload {
+    isolation_channel: Option<String>,
+}
+
+#[derive(Serialize)]
+struct CreatePayload {
+    name: String,
+    isolation_channel: Option<String>,
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let mutexbot = MutexBot::init().context("Failure to initialize action parameters")?;
+    let mut state = State::new(&mutexbot.api_key).context("Failure to initialize action state")?;
+
+    let isolation_channel = if mutexbot.isolation_channel.is_empty() {
+        None
+    } else {
+        Some(mutexbot.isolation_channel)
+    };
+
+    match mutexbot.mode.as_str() {
+        "reserve" => {
+            let notes = format!(
+                "Reserved by {}/{}/actions/runs/{}",
+                env::var("GITHUB_SERVER_URL")
+                    .context("Failure to load GITHUB_SERVER_URL env var")?,
+                env::var("GITHUB_REPOSITORY")
+                    .context("Failure to load GITHUB_REPOSITORY env var")?,
+                env::var("GITHUB_RUN_ID").context("Failure to load GITHUB_RUN_ID env var")?,
+            );
+            let duration = if mutexbot.duration.is_empty() {
+                None
+            } else {
+                Some(mutexbot.duration)
+            };
+            let payload = ReservePayload {
+                notes,
+                duration,
+                isolation_channel: isolation_channel.clone(),
+            };
+            loop {
+                match state
+                    .http
+                    .post(format!(
+                        "https://mutexbot.com/api/resources/global/{}/reserve",
+                        mutexbot.resource_name
+                    ))
+                    .json(&payload)
+                    .send()
+                    .await
+                {
+                    Ok(resp) => match resp.status() {
+                        StatusCode::CREATED => {
+                            info!("Resource reserved successfully");
+                            break;
+                        }
+                        StatusCode::CONFLICT => {
+                            info!("Resource already reserved, retrying in 5 seconds...");
+                            sleep(Duration::from_secs(5)).await;
```

Comment:
Optional: should we move such constants (timeouts/URLs) to config variables?

Reply:
I've moved the two different sleep durations (5s when waiting for a busy resource, 2s when retrying after a failure) to consts near further up, is that what you meant? Or does config variables mean that we convert them into inputs?

The API endpoints have moved into the `Mode` enum that I created to resolve the other comment.